contrib/selinux/syslog_ng.sh

```diff
@@ -1,63 +1,272 @@
+	# thus it needs to be removed (provided that the install path is not "/" itself)
+	sed -e 's:^\(.\+\)/$:\1:'
+}
+
+
+build_module() {
+	echo "Building and Loading Policy"
+	make -f /usr/share/selinux/devel/Makefile syslog_ng.pp || exit 1
+}
+
 
+add_ports() {
+	for entry in ${*}; do
+		port=${entry%/*}
+		proto=${entry#*/}
+		semanage port -a -t syslogd_port_t -p ${proto} ${port} 2>/dev/null || \
+		semanage port -m -t syslogd_port_t -p ${proto} ${port} 2>/dev/null
+	done
+}
 
-echo "Building and Loading Policy"
-make -f /usr/share/selinux/devel/Makefile
-/usr/sbin/semodule -i syslog_ng.pp
 
-# Fixing the file context on /opt/syslog-ng/sbin/syslog-ng
-/sbin/restorecon -F -R -v /opt/syslog-ng/sbin/syslog-ng
+install_module() {
+	/usr/sbin/semodule -i syslog_ng.pp -v || exit 1
+
+	# set up syslog-ng specific ports
+	PORTS=
+	for port in ${SYSLOG_NG_TCP_PORTS}; do PORTS="${PORTS} ${port}/tcp"; done
+	for port in ${SYSLOG_NG_UDP_PORTS}; do PORTS="${PORTS} ${port}/udp"; done
+	add_ports "${PORTS}"
+	
+	# Fixing the file context
+	/sbin/restorecon -F -Rv "${INSTALL_PATH}"
+	/sbin/restorecon -F -Rv /etc/init.d/syslog-ng
+	/sbin/restorecon -F -Rv /etc/rc.d/init.d/syslog-ng
+	/sbin/restorecon -F -Rv /dev/log
+	
+	echo -e "\nPlease restart syslog-ng. You can find more information about this in the README file."
+}
+
+remove_ports() {
+	for entry in ${*}; do
+		port=${entry%/*}
+		proto=${entry#*/}
+		semanage port -d -t syslogd_port_t -p ${proto} ${port} 2>/dev/null
+	done
+}
+
+remove_module() {
+	if semodule -l | grep -q syslog_ng; then
+		echo -n "Removing Syslog-ng SELinux policy module... "
+		
+		semodule --remove=syslog_ng
+		
+		# unconfigure syslog-ng specific ports
+		PORTS=
+		for port in ${SYSLOG_NG_TCP_PORTS}; do PORTS="${PORTS} ${port}/tcp"; done
+		for port in ${SYSLOG_NG_UDP_PORTS}; do PORTS="${PORTS} ${port}/udp"; done
+		remove_ports "${PORTS}"
+		
+		[ -f syslog_ng.pp ] && rm -f syslog_ng.pp
+		[ -f syslog_ng.te ] && rm -f syslog_ng.te
+		[ -f syslog_ng.fc ] && rm -f syslog_ng.fc
+		[ -f syslog_ng.if ] && rm -f syslog_ng.if
+		[ -d tmp ] && rm -Rf tmp
+		
+		echo "done."
+	else
+		echo "No installed Syslog-ng SELinux policy module was found. Nothing to do..."
+	fi
+}
+
+DIRNAME=$( dirname ${0} )
+cd ${DIRNAME}
+USAGE="$0\t[ --install-dir <DIRECTORY> | --remove | --help ]\n\n$0:\tA tool for building and managing the SELinux policy for the\n\t\tdefault syslog-ng installation."
+
+while [ -n "${1}" ]; do
+	case "${1}" in
+		--install-dir)
+			check_dir "${2}" || exit 1
+			INPUT="${2}"
+			shift 2
+			;;
+		--remove)
+			[ "${TASK_SELECTED}" = "install" ] && TASK_SELECTED="remove"
+			shift
+			;;
+		--help)
+			# if --help is supplied, the help message will be printed independently of any other options being specified
+			TASK_SELECTED="showhelp"
+			shift
+			;;
+		*)
+			echo -e "ERROR: Invalid option: ${1}\n${USAGE}" >&2
+			exit 1
+			;;
+	esac
+
+done
+
+case "${TASK_SELECTED}" in
```

Comment:
The primary reason why I designed the code this way is that:
- I wanted to make sure, that all CLI arguments are parsed first
- some arguments take precedence over others: help > remove > install
- I wanted to avoid the script acting differently when the arguments are specified in a different order
- invoking functionality from the first switch would cause the functionality being invoked multiple times, in case an argument were to be written multiple times

This seemed to be the easiest and most efficient way to do it. (Well, at least for me.)

Your other comments are perfectly valid, and I have already corrected them.